Add Storage.store and ability to store to the web

rschamp opened the pull request:
### Resolves

_What Github issue does this resolve (please include link)?_
Towards https://github.com/LLK/scratch-gui/issues/2968

### Proposed Changes

_Describe what this Pull Request does_
- Remove LocalHelper
- Add SB3 DataFormat
- Replace fetch-only "source" concept with 2-way "store" concept
- Enhance `.addWebStore` so we can make `PUT`/`POST` requests to web stores
- Add `.store` method

### Reason for Changes

_Explain why these changes should be made_
We need to be able to modify projects on the project server using storage. This is a step towards that goal.

Commit:
refactor: remove LocalHelper

`LocalHelper` is unused currently, so remove it for simplicity

BREAKING CHANGE: The private `localHelper` property has been removed

Commit:
fix: assume BuiltinHelper ids to be existy, not truthy

BuiltinHelper was assuming the default project (id 0) was not cached, when it was

Commit:
feat(data format): add SB3 DataFormat, so we can save SB3s!

Commit:
fix: increase strictness of BuiltinHelper

BuiltinHelper should not cache ids of `undefined` or `null`.

BREAKING CHANGE: Began throwing errors when attempting to use `BuiltinHelper.cache` with non-existy ids

Commit:
feat: add logger

Use minilog for runtime deprecation warnings

Commit:
feat: replace "sources" with "stores"

Sources are only for fetching, but stores allow fetching/creating/updating.  Add deprecation warnings when attempting to access "sources".  addStore accepts two additional functions, for creation and updates. Additionally, the functions can return strings or an object if additional configuration to the request to the URL is necessary. The attributes of the object are passed as options to `nets`.

BREAKING CHANGE: WebHelper.sources has been replaced by WebHelper.stores. WebHelper.addSource has been replaced by WebHelper.addStore. Storage.addWebSource has been replaced by Storage.addWebStore. BuiltinHelper.cache has been normalized to BuiltinHelper.store.

Commit:
feat: add Storage.store

Storage.store attempts an update or create on the webHelper first, and then supplies the resulting id to BuiltinHelper to cache the result.

Commit:
refactor: replace "fulfill" with "resolve"

Only necessary for consistency. And fulfill was bugging me since resolve is the conventional term for this function.

kchadha commented on src/BuiltinHelper.js:
```diff
@@ -93,19 +95,35 @@ class BuiltinHelper extends Helper {
+    store (assetType, dataFormat, data, id) {
         if (!dataFormat) throw new Error('Data cached without specifying its format');
-        if (id) {
+        if (id !== null && typeof id !== 'undefined') {
```
Do we want to allow empty string IDs?

cwillisf replied:
+1, especially considering that the empty string is converted to an "Untitled #"-type title in Scratch 2.0

kchadha commented on src/ScratchStorage.js:
```diff
@@ -171,6 +183,27 @@ class ScratchStorage {
+     * Store an asset by type & ID.
+     * @param {AssetType} assetType - The type of asset to fetch. This also determines which asset store to use.
+     * @param {DataFormat} [dataFormat] - Optional: load this format instead of the AssetType's default.
+     * @param {DataFormat} [data] - Optional: store this data, rather than the data in the asset
```
Is this type annotation incorrect? Also is `data` actually optional?

kchadha commented on src/ScratchStorage.js:
```diff
@@ -171,6 +183,27 @@ class ScratchStorage {
+     * @param {DataFormat} [dataFormat] - Optional: load this format instead of the AssetType's default.
+     * @param {DataFormat} [data] - Optional: store this data, rather than the data in the asset
+     * @param {?string} assetId - The ID of the asset to fetch: a project ID, MD5, etc.
+     * @return {Promise.<id, title>} A promise for asset metadata
```
Is this type annotation correct? I thought `.<x,y>` was used to denote instances of parameterized type (e.g. for maps it would be the type of the keys and the values respectively)...

cwillisf replied:
I found this example on the JSDoc web site: `@return {Promise<string>}`

I think the proper way to do this might be something like: `@return {Promise<{id: string, title: string}>}` (or is the `id` a `Number`?)

rschamp replied:
The stuff in the returned object could be anything the web store returns so I think `Promise.<object>` is what we want.

kchadha commented on src/ScratchStorage.js:
```diff
@@ -171,6 +183,27 @@ class ScratchStorage {
+                this.webHelper.store(assetType, dataFormat, data, assetId)
+                    .then(body => {
+                        this.builtinHelper.store(assetType, dataFormat, data, body.id);
+                        return resolve(body);
```
Is the `return` unnecessary here?

rschamp replied:
By convention you should `return` when you call the callback so that you don't accidentally do stuff after it (like calling the callback a second time). Not strictly necessary in terms of code though.

kchadha commented on src/WebHelper.js:
```diff
@@ -45,52 +65,110 @@ class WebHelper extends Helper {
+    store (assetType, dataFormat, data, assetId) {
+        const asset = new Asset(assetType, assetId, dataFormat);
+        // If we have an asset id, we should update, otherwise create to get an id
+        const create = assetId === null || typeof assetId === 'undefined';
```
Same question here about empty string assetId...

kchadha commented on src/WebHelper.js:
```diff
@@ -45,52 +65,110 @@ class WebHelper extends Helper {
+        // If we have an asset id, we should update, otherwise create to get an id
+        const create = assetId === null || typeof assetId === 'undefined';
+
+        // Use the first source with the appropriate asset type and url function
```
source -> store?

cwillisf commented on src/WebHelper.js:
```diff
@@ -1,36 +1,56 @@
  * @param {Asset} - The asset for which the URL should be computed.
- * @returns {string} - The URL for the asset.
+ * @returns {(string|object)} - A string representing the URL for the asset request OR an object with configuration for
+ *                              the underlying `nets` call (necessary for configuring e.g. authentication)
```
I'm OK with this for now, but it seems like we might need a `nets`-independent way to handle this kind of thing in the future.

rschamp replied:
Agreed. Could be complicated though.

rschamp replied:
E.g., here's how we have to do it atm (and the reason for this change):
```js
    getProjectCreateConfig () {
        return {
            url: `${this.projectHost}/`,
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded'
            },
            withCredentials: true,
            json: true
        };
```

kchadha commented on src/WebHelper.js:
```diff
@@ -45,52 +65,110 @@ class WebHelper extends Helper {
+                reqConfig = {
+                    url: reqConfig
+                };
+            }
```
You should probably have an `else { return reject(...); }` here...

rschamp replied:
This is for backwards compatibility from when we only returned strings for urls. So now it converts those results into objects, setting the url attribute to that result.

kchadha reviewed (commented):
In general, LGTM. I have a few nits and questions in the comments.

cwillisf reviewed (commented):
Similar to @kchadha, I think this looks pretty good with a few nits & questions.

Commit:
fix: use Object.assign instead of spread operator

The webpack config doesn't support the spread operator yet (even though gui's does)

Commit:
fix: address review feedback

Thanks @kchadha and @cwillisf!
